### ISSUE TITLE
voctree doesn't compile due to include paths

### ISSUE THREAD

**Issue by zvrba** [poparteu/openMVG#246]
When compiling current `popart_develop` I get the following error when compiling `createVoctree.cpp`:

```
In file included from /media/install/SOURCE/openMVG.git/src/software/Localization/createVoctree.cpp:4:0:
/media/install/INSTALL/include/openMVG/voctree/descriptor_loader.hpp:94:33: fatal error: descriptor_loader.tcc: No such file or directory
 #include "descriptor_loader.tcc"
                                 ^
compilation terminated.
```

The problem is that `.hpp` file includes `.tcc` file. I tried to modify `CMakeLists.txt` in the voctree directory by adding

```
SET(CMAKE_INCLUDE_CURRENT_DIR ON)
file(
  GLOB_RECURSE
  voctree_header
  *.hpp *.tcc # .tcc isn't mentioned here
)
```

to no effect, i.e., the compilation still fails.

**Comment by zvrba** [comment]
Adding also `SET(CMAKE_INCLUDE_CURRENT_DIR_IN_INTERFACE ON)` fixes the above problem, but now I get the following:

```
In file included from /media/install/INSTALL/include/openMVG/voctree/descriptor_loader.hpp:94:0,
                 from /media/install/SOURCE/openMVG.git/src/software/Localization/voctreeQueryUtility.cpp:5:
/media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.tcc:15:8: error: redefinition of ‘template<class DescriptorT, class FileDescriptorT> std::size_t openMVG::voctree::readDescFromFiles(const string&, std::vector<_RealType>&, std::vector<long unsigned int>&)’
 size_t readDescFromFiles(const std::string &fileFullPath, std::vector<DescriptorT>& descriptors, std::vector<size_t> &numFeatures)
        ^
In file included from /media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.hpp:94:0,
                 from /media/install/SOURCE/openMVG.git/src/openMVG/voctree/databaseIO.tcc:1,
                 from /media/install/INSTALL/include/openMVG/voctree/databaseIO.hpp:98,
                 from /media/install/SOURCE/openMVG.git/src/software/Localization/voctreeQueryUtility.cpp:3:
/media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.tcc:15:8: note: ‘template<class DescriptorT, class FileDescriptorT> std::size_t openMVG::voctree::readDescFromFiles(const string&, std::vector<_RealType>&, std::vector<long unsigned int>&)’ previously declared here
 size_t readDescFromFiles(const std::string &fileFullPath, std::vector<DescriptorT>& descriptors, std::vector<size_t> &numFeatures)
        ^
```

I think the simplest fix is to remove the `.tcc` files and put the definition in the `.hpp`. There's no disadvantage to doing so since templates cannot be compiled to `.o` files anyway.

**Comment by simogasp** [comment]
It's strange I never had this problem and this part of the code hasn't changed in a long time.
For sure there are some stuff to fix:
- the headers do not need to be added in `ADD_LIBRARY`
- we forgot about the tcc and it should be also installed when we do `make install` which is not the case at the moment

That being said, it's bizarre that it finds 
```
#include "database.hpp"
#include "vocabulary_tree.hpp"
```
and not `#include "descriptor_loader.tcc"`.

Just to be sure, have you tried to remove CMakeCache and redo cmake?

**Comment by zvrba** [comment]
Not doing `rm -rf build` after checking out another branch was the first mistake I made. I think the simplest solution is to just merge the tcc file into the hpp files.

I think that the (non-)existence of the problem depends on the cmake version because it's related to interface include directoried. Here I use cmake 3.0.2.

**Comment by simogasp** [comment]
So is it working now without all the changes you made?
I'm using cmake 3.6 on OSX, the travis version has been changed to 3.4 in the other branch, i have another ubuntu with cmake 3.1 and the issue never came up.

About tcc, i don't know, i did it like that because i like to have "clean" headers, but if it is a problem we can merge them, after all the rest of the lib do not use tcc.

**Comment by zvrba** [comment]
I'm compiling with clang now, that's the first (and only) time I compiled it with w/o trouble.

**Comment by simogasp** [comment]
can you `make VERBOSE=1` and copy paste the compilation line that generate the problem?

**Comment by simogasp** [comment]
Also, as a side note, the `file(GLOB_RECURSE ...` is not a good way to do the list of file, it's better to do it explicitly because of the cmake cache: ie even if you change the pattern in `file` you are better off by deleting CMakeCache or `rm -rf build` as cmake do not re-do the listing of file. In a future refactoring of the lib we need to replace all these with explicit lists of files.

**Comment by zvrba** [comment]
Here's the full output:

```
cd /media/install/SOURCE/openMVG.git/build/software/Localization && /usr/bin/c++   -DBOOST_NO_CXX11_SCOPED_ENUMS -DEIGEN_MPL2_ONLY -DHAVE_ALEMBIC -DHAVE_BOOST -DHAVE_CXX11_CHRONO -DHAVE_CXX11_THREAD -DHAVE_FBX -DHAVE_OPENCV -DOPENMVG_USE_CXX11 -DOPENMVG_USE_OPENMP -DOPENMVG_USE_SSE -DWANTS_POPART_COUT -march=native -Werror=return-type -Werror=switch -Werror=return-local-addr --std=c++11 -fopenmp -O3 -I/media/install/SOURCE/openMVG.git/src/third_party/eigen -I/media/install/INSTALL/include/opencv -I/media/install/INSTALL/include -I/media/install/INSTALL/include/OpenEXR -I/media/install/SOURCE/openMVG.git/src -I/usr/include/x86_64-linux-gnu -I/media/install/SOURCE/openMVG.git/src/third_party -I/media/install/SOURCE/openMVG.git/src/dependencies -I/media/install/SOURCE/openMVG.git/src/dependencies/cereal/include -I/media/install/SOURCE/openMVG.git/src/third_party/lemon -I/media/install/SOURCE/openMVG.git/build/third_party/lemon -I/media/install/SOURCE/openMVG.git/src/third_party/ceres-solver/include -I/media/install/SOURCE/openMVG.git/src/third_party/ceres-solver/internal/ceres/miniglog -I/media/install/SOURCE/openMVG.git/build/third_party/ceres-solver/config -I/media/install/SOURCE/openMVG.git/src/third_party/flann/src/cpp -I/media/install/SOURCE/openMVG.git/src/dependencies/osi_clp/CoinUtils/src -I/media/install/SOURCE/openMVG.git/src/dependencies/osi_clp/Clp/src -I/media/install/SOURCE/openMVG.git/src/dependencies/osi_clp/Clp/src/OsiClp -I/media/install/SOURCE/openMVG.git/src/dependencies/osi_clp/Osi/src/Osi -I/media/install/SOURCE/openMVG.git/build/openMVG/voctree -I/media/install/SOURCE/openMVG.git/src/openMVG/voctree    -o CMakeFiles/openMVG_main_voctreeQueryUtility.dir/voctreeQueryUtility.cpp.o -c /media/install/SOURCE/openMVG.git/src/software/Localization/voctreeQueryUtility.cpp
In file included from /media/install/INSTALL/include/openMVG/voctree/descriptor_loader.hpp:94:0,
                 from /media/install/SOURCE/openMVG.git/src/software/Localization/voctreeQueryUtility.cpp:5:
/media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.tcc:15:8: error: redefinition of ‘template<class DescriptorT, class FileDescriptorT> std::size_t openMVG::voctree::readDescFromFiles(const string&, std::vector<_RealType>&, std::vector<long unsigned int>&)’
 size_t readDescFromFiles(const std::string &fileFullPath, std::vector<DescriptorT>& descriptors, std::vector<size_t> &numFeatures)
        ^
In file included from /media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.hpp:94:0,
                 from /media/install/SOURCE/openMVG.git/src/openMVG/voctree/databaseIO.tcc:1,
                 from /media/install/INSTALL/include/openMVG/voctree/databaseIO.hpp:98,
                 from /media/install/SOURCE/openMVG.git/src/software/Localization/voctreeQueryUtility.cpp:3:
/media/install/SOURCE/openMVG.git/src/openMVG/voctree/descriptor_loader.tcc:15:8: note: ‘template<class DescriptorT, class FileDescriptorT> std::size_t openMVG::voctree::readDescFromFiles(const string&, std::vector<_RealType>&, std::vector<long unsigned int>&)’ previously declared here
 size_t readDescFromFiles(const std::string &fileFullPath, std::vector<DescriptorT>& descriptors, std::vector<size_t> &numFeatures)
        ^
```

**Comment by simogasp** [comment]
Ok i knew it was something like that! :-D
```
In file included from /media/install/INSTALL/include/openMVG/voctree/descriptor_loader.hpp
```
You have an older version of openMVG installed in `/media/install/INSTALL/include`, don't you?
So since you have a `-I/media/install/INSTALL/include` in the line command it is picking the headers from there. And of course tcc is not installed, as I said before. That has to be correct though. But I need to remove the older version from the install path if you want to compile properly.
I happened to me quite few times, and it is painful to discover that! ;-)

**Comment by zvrba** [comment]
Yes, that's right, another version is installed there. But I've fixed the current build by commenting out `#include <.../descriptor_loader.hpp>` in voctreeQueryUtility.cpp and voctreeStatistics.cpp.

**Comment by simogasp** [comment]
That's another can of worms, we need to remove the useless includes... :-(
Anyway be careful compiling using the headers from another version, as sadly there is a lot of cpp/template code inside the headers, so even if it compiles fine it may end up with unexpected behaviours.

**Comment by simogasp** [comment]
> in-tree includes (.../SOURCE) should have precedence over "system" includes (..../INSTALL).

It depends on where that `-I/media/install/INSTALL/include` is coming from. If it is because another dependency is installed there (eg. cctag?) there is not much to do other than remove the older version. I don't think it comes from the build system. Then I agree with you that we need to do better all the includes using the in-tree and the install when necessary using `$<BUILD_INTERFACE` and `$<INSTALL_INTERFACE`. But again this is going to be fixed with that planned general (huge) revision

**Comment by simogasp** [comment]
tcc problem is fixed by poparteu/openMVG#247

**Comment by zvrba** [comment]
Yes, it comes from another dependency (Alembic). But include directories are searched in order they are specified, so all `.../SOURCE` directories should come before any `.../INSTALL` directory.
